File: app/src/test/java/com/github/vase4kin/teamcityapp/overview/data/BuildDetailsImplTest.java
/*
 * Copyright 2019 Andrey Tolpeev
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.github.vase4kin.teamcityapp.overview.data;

import com.github.vase4kin.teamcityapp.agents.api.Agent;
import com.github.vase4kin.teamcityapp.artifact.api.Artifacts;
import com.github.vase4kin.teamcityapp.buildlist.api.Build;
import com.github.vase4kin.teamcityapp.buildlist.api.CanceledInfo;
import com.github.vase4kin.teamcityapp.buildlist.api.Triggered;
import com.github.vase4kin.teamcityapp.buildlist.api.User;
import com.github.vase4kin.teamcityapp.changes.api.Changes;
import com.github.vase4kin.teamcityapp.navigation.api.BuildType;
import com.github.vase4kin.teamcityapp.properties.api.Properties;
import com.github.vase4kin.teamcityapp.tests.api.TestOccurrences;
import com.github.vase4kin.teamcityapp.utils.IconUtils;

import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import static com.github.vase4kin.teamcityapp.overview.data.BuildDetails.STATE_QUEUED;
import static com.github.vase4kin.teamcityapp.overview.data.BuildDetails.STATE_RUNNING;
import static com.github.vase4kin.teamcityapp.overview.data.BuildDetails.STATUS_FAILURE;
import static com.github.vase4kin.teamcityapp.overview.data.BuildDetails.STATUS_SUCCESS;
import static com.github.vase4kin.teamcityapp.overview.data.BuildDetailsImpl.TEXT_NO_NUMBER;
import static com.github.vase4kin.teamcityapp.overview.data.BuildDetailsImpl.TEXT_QUEUED_BUILD;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.equalTo;
import static org.hamcrest.Matchers.is;
import static org.mockito.Mockito.when;

public class BuildDetailsImplTest {

    /**
     * Test data
     */
    private static final String TIME_STAMP = "20160621T233008+0700";
    private static final String TIME_FORMATTED = "21 Jun 16 23:30";

    @Mock
    private Artifacts mArtifacts;

    @Mock
    private TestOccurrences mTestOccurrences;

    @Mock
    private Changes mChanges;

    @Mock
    private Properties mProperties;

    @Mock
    private Triggered mTriggered;

    @Mock
    private BuildType mBuildType;


    @Mock
    private User mUser;

    @Mock
    private CanceledInfo mCanceledInfo;

    @Mock
    private Build mBuild;
    private Agent mAgent = new Agent("name");

    private BuildDetails mBuildDetails;

    @Before
    public void setUp() {
        MockitoAnnotations.initMocks(this);
        mBuildDetails = new BuildDetailsImpl(mBuild);
    }

    @Test
    public void testGetHref() {
        when(mBuild.getHref()).thenReturn("href");
        assertThat(mBuildDetails.getHref(), is(equalTo("href")));
    }

    @Test
    public void testGetWebUrl() {
        when(mBuild.getWebUrl()).thenReturn("webUrl");
        assertThat(mBuildDetails.getWebUrl(), is(equalTo("webUrl")));
    }

    @Test
    public void testGetChangesHref() {
        when(mBuild.getChanges()).thenReturn(mChanges);
        when(mChanges.getHref()).thenReturn("href");
        assertThat(mBuildDetails.getChangesHref(), is(equalTo("href")));
    }

    @Test
    public void testGetStatusTextIfBuildIsNotQueued() {
        when(mBuild.getStatusText()).thenReturn("status");
        when(mBuild.getState()).thenReturn("state");
        assertThat(mBuildDetails.getStatusText(), is(equalTo("status")));
    }

    @Test
    public void testGetStatusTextIfBuildIsQueuedButDoesNotHaveWaitReason() {
        when(mBuild.getState()).thenReturn(STATE_QUEUED);
        assertThat(mBuildDetails.getStatusText(), is(equalTo(TEXT_QUEUED_BUILD)));
    }

    @Test
    public void testGetStatusTextIfBuildIsQueuedButHasWaitReason() {
        when(mBuild.getState()).thenReturn(STATE_QUEUED);
        when(mBuild.getWaitReason()).thenReturn("waitReason");
        assertThat(mBuildDetails.getStatusText(), is(equalTo("waitReason")));
    }

    @Test
    public void testGetStatusIcon() {
        when(mBuild.getState()).thenReturn("");
        when(mBuild.getStatus()).thenReturn("");
        assertThat(mBuildDetails.getStatusIcon(), is(equalTo(IconUtils.ICON_SUCCESS)));
    }

    @Test
    public void testIsRunning() {
        when(mBuild.getState()).thenReturn(STATE_RUNNING);
        assertThat(mBuildDetails.isRunning(), is(equalTo(true)));
    }

    @Test
    public void testIsQueued() {
        when(mBuild.getState()).thenReturn(STATE_QUEUED);
        assertThat(mBuildDetails.isQueued(), is(equalTo(true)));
    }

    @Test
    public void testIsSuccess() {
        when(mBuild.getStatus()).thenReturn(STATUS_SUCCESS);
        assertThat(mBuildDetails.isSuccess(), is(equalTo(true)));
    }

    @Test
    public void testIsFailed() {
        when(mBuild.getStatus()).thenReturn(STATUS_FAILURE);
        assertThat(mBuildDetails.isFailed(), is(equalTo(true)));
    }

    @Test
    public void testHasCancellationInfo() {
        when(mBuild.getCanceledInfo()).thenReturn(mCanceledInfo);
        assertThat(mBuildDetails.hasCancellationInfo(), is(equalTo(true)));
    }

    @Test
    public void testHasUserInfoWhoCancelledBuild() {
        when(mBuild.getCanceledInfo()).thenReturn(mCanceledInfo);
        when(mCanceledInfo.getUser()).thenReturn(mUser);
        assertThat(mBuildDetails.hasUserInfoWhoCancelledBuild(), is(equalTo(true)));
    }

    @Test
    public void testGetUserNameWhoCancelledBuild() {
        when(mBuild.getCanceledInfo()).thenReturn(mCanceledInfo);
        when(mCanceledInfo.getUser()).thenReturn(mUser);
        when(mUser.getName()).thenReturn("user");
        assertThat(mBuildDetails.getUserNameWhoCancelledBuild(), is(equalTo("user")));
    }

    @Test
    public void testGetCancellationTime() {
        when(mBuild.getCanceledInfo()).thenReturn(mCanceledInfo);
        when(mCanceledInfo.getTimestamp()).thenReturn(TIME_STAMP);
        assertThat(mBuildDetails.getCancellationTime(), is(equalTo(TIME_FORMATTED)));
    }

    @Test
    public void testGetStartDate() {
        when(mBuild.getStartDate()).thenReturn(TIME_STAMP);
        assertThat(mBuildDetails.getStartDate(), is(equalTo(TIME_FORMATTED)));
    }

    @Test
    public void testGetStartDateFormattedAsHeader() {
        when(mBuild.getStartDate()).thenReturn(TIME_STAMP);
        assertThat(mBuildDetails.getStartDateFormattedAsHeader(), is(equalTo("21 June")));
    }

    @Test
    public void testGetQueuedDate() {
        when(mBuild.getQueuedDate()).thenReturn(TIME_STAMP);
        assertThat(mBuildDetails.getQueuedDate(), is(equalTo(TIME_FORMATTED)));
    }

    @Test
    public void testGetFinishTime() {
        when(mBuild.getStartDate()).thenReturn(TIME_STAMP);
        when(mBuild.getFinishDate()).thenReturn(TIME_STAMP);
        assertThat(mBuildDetails.getFinishTime(), is(equalTo("21 Jun 16 23:30 - 23:30 (0s)")));
    }

    @Test
    public void testGetEstimatedStartTime() {
        when(mBuild.getStartEstimate()).thenReturn(TIME_STAMP);
        assertThat(mBuildDetails.getEstimatedStartTime(), is(equalTo(TIME_FORMATTED)));
    }

    @Test
    public void testGetBranchName() {
        when(mBuild.getBranchName()).thenReturn("branch");
        assertThat(mBuildDetails.getBranchName(), is(equalTo("branch")));
    }

    @Test
    public void testHasAgentInfo() {
        when(mBuild.getAgent()).thenReturn(mAgent);
        assertThat(mBuildDetails.hasAgentInfo(), is(equalTo(true)));
    }

    @Test
    public void testGetAgentName() {
        when(mBuild.getAgent()).thenReturn(mAgent);
        assertThat(mBuildDetails.getAgentName(), is(equalTo("name")));
    }

    @Test
    public void testIsTriggeredByVcs() {
        when(mBuild.getTriggered()).thenReturn(mTriggered);
        when(mTriggered.isVcs()).thenReturn(true);
        assertThat(mBuildDetails.isTriggeredByVcs(), is(equalTo(true)));
    }

    @Test
    public void testIsTriggeredByUser() {
        when(mBuild.getTriggered()).thenReturn(mTriggered);
        when(mTriggered.isUser()).thenReturn(true);
        assertThat(mBuildDetails.isTriggeredByUser(), is(equalTo(true)));
    }

    @Test
    public void testIsRestarted() {
        when(mBuild.getTriggered()).thenReturn(mTriggered);
        when(mTriggered.isRestarted()).thenReturn(true);
        assertThat(mBuildDetails.isRestarted(), is(equalTo(true)));
    }

    @Test
    public void testIsTriggeredByBuildType() {
        when(mBuild.getTriggered()).thenReturn(mTriggered);
        when(mTriggered.isBuildType()).thenReturn(true);
        assertThat(mBuildDetails.isTriggeredByBuildType(), is(equalTo(true)));
    }

    @Test
    public void testIsTriggeredByUnknown() {
        when(mBuild.getTriggered()).thenReturn(mTriggered);
        when(mTriggered.isUnknown()).thenReturn(true);
        assertThat(mBuildDetails.isTriggeredByUnknown(), is(equalTo(true)));
    }

    @Test
    public void tetsGetTriggeredDetails() {
        when(mBuild.getTriggered()).thenReturn(mTriggered);
        when(mTriggered.getDetails()).thenReturn("details");
        assertThat(mBuildDetails.getTriggeredDetails(), is(equalTo("details")));
    }

    @Test
    public void testGetUserNameOfUserWhoTriggeredBuild() {
        when(mBuild.getTriggered()).thenReturn(mTriggered);
        when(mTriggered.getUser()).thenReturn(mUser);
        when(mUser.getName()).thenReturn("name");
        assertThat(mBuildDetails.getUserNameOfUserWhoTriggeredBuild(), is(equalTo("name")));
    }

    @Test
    public void testGetNameOfTriggeredBuildTypeIfBuildTypeIsDeleted() {
        when(mBuild.getTriggered()).thenReturn(mTriggered);
        when(mTriggered.getBuildType()).thenReturn(null);
        assertThat(mBuildDetails.getNameOfTriggeredBuildType(), is(equalTo(BuildDetailsImpl.TEXT_DELETED_CONFIGURATION)));
    }

    @Test
    public void testGetNameOfTriggeredBuildType() {
        when(mBuild.getTriggered()).thenReturn(mTriggered);
        when(mTriggered.getBuildType()).thenReturn(mBuildType);
        when(mBuildType.getProjectName()).thenReturn("pn");
        when(mBuildType.getName()).thenReturn("cn");
        assertThat(mBuildDetails.getNameOfTriggeredBuildType(), is(equalTo("pn cn")));
    }

    @Test
    public void testGetBuildTypeId() {
        when(mBuild.getBuildTypeId()).thenReturn("id");
        assertThat(mBuildDetails.getBuildTypeId(), is(equalTo("id")));
    }

    @Test
    public void testHasBuildTypeInfo() {
        when(mBuild.getBuildType()).thenReturn(null);
        assertThat(mBuildDetails.hasBuildTypeInfo(), is(equalTo(false)));
    }

    @Test
    public void testGetBuildTypeFullName() {
        when(mBuild.getBuildType()).thenReturn(mBuildType);
        when(mBuildType.getProjectName()).thenReturn("project");
        when(mBuildType.getName()).thenReturn("name");
        assertThat(mBuildDetails.getBuildTypeFullName(), is(equalTo("project - name")));
    }

    @Test
    public void testGetBuildTypeName() {
        when(mBuild.getBuildType()).thenReturn(mBuildType);
        when(mBuildType.getName()).thenReturn("name");
        assertThat(mBuildDetails.getBuildTypeName(), is(equalTo("name")));
    }

    @Test
    public void testGetProjectName() {
        when(mBuild.getBuildType()).thenReturn(mBuildType);
        when(mBuildType.getProjectName()).thenReturn("project");
        assertThat(mBuildDetails.getProjectName(), is(equalTo("project")));
    }

    @Test
    public void testGetProjectId() {
        when(mBuild.getBuildType()).thenReturn(mBuildType);
        when(mBuildType.getProjectId()).thenReturn("id");
        assertThat(mBuildDetails.getProjectId(), is(equalTo("id")));
    }

    @Test
    public void testGetBuildId() {
        when(mBuild.getId()).thenReturn("id");
        assertThat(mBuildDetails.getId(), is(equalTo("id")));
    }

    @Test
    public void testGetProperties() {
        when(mBuild.getProperties()).thenReturn(mProperties);
        assertThat(mBuildDetails.getProperties(), is(equalTo(mProperties)));
    }

    @Test
    public void testIsTriggeredByUserIfTriggeredIsNull() {
        when(mBuild.getTriggered()).thenReturn(null);
        assertThat(mBuildDetails.isTriggeredByUser("user"), is(equalTo(false)));
    }

    @Test
    public void testIsTriggeredByUserIfTriggeredByNotUser() {
        when(mBuild.getTriggered()).thenReturn(mTriggered);
        when(mTriggered.isUser()).thenReturn(false);
        assertThat(mBuildDetails.isTriggeredByUser("user"), is(equalTo(false)));
    }

    @Test
    public void testIsTriggeredByUserIfUserIsNull() {
        when(mBuild.getTriggered()).thenReturn(mTriggered);
        when(mTriggered.isUser()).thenReturn(true);
        when(mTriggered.getUser()).thenReturn(null);
        assertThat(mBuildDetails.isTriggeredByUser("user"), is(equalTo(false)));
    }

    @Test
    public void testIsTriggeredByUserIfUserIsNotNull() {
        when(mBuild.getTriggered()).thenReturn(mTriggered);
        when(mTriggered.isUser()).thenReturn(true);
        when(mTriggered.getUser()).thenReturn(new User("user", null));
        assertThat(mBuildDetails.isTriggeredByUser("user"), is(equalTo(true)));
    }

    @Test
    public void testHasTests() {
        when(mBuild.getTestOccurrences()).thenReturn(mTestOccurrences);
        assertThat(mBuildDetails.hasTests(), is(equalTo(true)));
    }

    @Test
    public void testGetTestsHref() {
        when(mBuild.getTestOccurrences()).thenReturn(mTestOccurrences);
        when(mTestOccurrences.getHref()).thenReturn("href");
        assertThat(mBuildDetails.getTestsHref(), is(equalTo("href")));
    }

    @Test
    public void testGetPassedTestsCount() {
        when(mBuild.getTestOccurrences()).thenReturn(mTestOccurrences);
        when(mTestOccurrences.getPassed()).thenReturn(3);
        assertThat(mBuildDetails.getPassedTestCount(), is(equalTo(3)));
    }

    @Test
    public void testGetFailedTestsCount() {
        when(mBuild.getTestOccurrences()).thenReturn(mTestOccurrences);
        when(mTestOccurrences.getFailed()).thenReturn(4);
        assertThat(mBuildDetails.getFailedTestCount(), is(equalTo(4)));
    }

    @Test
    public void testGetIgnoredTestsCount() {
        when(mBuild.getTestOccurrences()).thenReturn(mTestOccurrences);
        when(mTestOccurrences.getIgnored()).thenReturn(5);
        assertThat(mBuildDetails.getIgnoredTestCount(), is(equalTo(5)));
    }

    @Test
    public void testGetArtifactsHref() {
        when(mBuild.getArtifacts()).thenReturn(mArtifacts);
        when(mArtifacts.getHref()).thenReturn("href");
        assertThat(mBuildDetails.getArtifactsHref(), is(equalTo("href")));
    }

    @Test
    public void testGetNumber() {
        when(mBuild.getNumber()).thenReturn("Number");
        assertThat(mBuildDetails.getNumber(), is(equalTo("Number")));
    }

    @Test
    public void testGetNumberIfNumberIsNull() {
        when(mBuild.getNumber()).thenReturn(null);
        assertThat(mBuildDetails.getNumber(), is(equalTo(TEXT_NO_NUMBER)));
    }

    @Test
    public void testIsPersonal() {
        when(mBuild.isPersonal()).thenReturn(true);
        assertThat(mBuildDetails.isPersonal(), is(equalTo(true)));
    }

    @Test
    public void testIsPinned() {
        when(mBuild.isPinned()).thenReturn(true);
        assertThat(mBuildDetails.isPinned(), is(equalTo(true)));
    }

    @Test
    public void testToBuild() {
        assertThat(mBuildDetails.toBuild(), is(equalTo(mBuild)));
    }

}

File: app/src/test/java/com/github/vase4kin/teamcityapp/overview/data/BuildDetailsImplTest.kt
/*
 * Copyright 2019 Andrey Tolpeev
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.github.vase4kin.teamcityapp.overview.data

import com.github.vase4kin.teamcityapp.agents.api.Agent
import com.github.vase4kin.teamcityapp.artifact.api.Artifacts
import com.github.vase4kin.teamcityapp.buildlist.api.Build
import com.github.vase4kin.teamcityapp.buildlist.api.CanceledInfo
import com.github.vase4kin.teamcityapp.buildlist.api.Triggered
import com.github.vase4kin.teamcityapp.buildlist.api.User
import com.github.vase4kin.teamcityapp.changes.api.Changes
import com.github.vase4kin.teamcityapp.navigation.api.BuildType
import com.github.vase4kin.teamcityapp.overview.data.BuildDetails.Companion.STATE_QUEUED
import com.github.vase4kin.teamcityapp.overview.data.BuildDetails.Companion.STATE_RUNNING
import com.github.vase4kin.teamcityapp.overview.data.BuildDetails.Companion.STATUS_FAILURE
import com.github.vase4kin.teamcityapp.overview.data.BuildDetails.Companion.STATUS_SUCCESS
import com.github.vase4kin.teamcityapp.overview.data.BuildDetailsImpl.Companion.TEXT_NO_NUMBER
import com.github.vase4kin.teamcityapp.overview.data.BuildDetailsImpl.Companion.TEXT_QUEUED_BUILD
import com.github.vase4kin.teamcityapp.properties.api.Properties
import com.github.vase4kin.teamcityapp.tests.api.TestOccurrences
import com.github.vase4kin.teamcityapp.utils.IconUtils
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.`is`
import org.hamcrest.Matchers.equalTo
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mock
import org.mockito.Mockito.`when`
import org.mockito.runners.MockitoJUnitRunner

private const val TIME_STAMP = "20160621T233008+0700"
private const val TIME_FORMATTED = "21 Jun 16 23:30"

@RunWith(MockitoJUnitRunner::class)
class BuildDetailsImplTest {

    @Mock
    private lateinit var artifacts: Artifacts
    @Mock
    private lateinit var testOccurrences: TestOccurrences
    @Mock
    private lateinit var changes: Changes
    @Mock
    private lateinit var properties: Properties
    @Mock
    private lateinit var triggered: Triggered
    @Mock
    private lateinit var buildType: BuildType
    @Mock
    private lateinit var user: User
    @Mock
    private lateinit var canceledInfo: CanceledInfo
    @Mock
    private lateinit var build: Build
    private val mAgent = Agent("name")
    private lateinit var mBuildDetails: BuildDetails

    @Before
    fun setUp() {
        mBuildDetails = BuildDetailsImpl(build)
    }

    @Test
    fun testGetHref() {
        `when`(build.href).thenReturn("href")
        assertThat(mBuildDetails.href, `is`(equalTo("href")))
    }

    @Test
    fun testGetWebUrl() {
        `when`(build.webUrl).thenReturn("webUrl")
        assertThat(mBuildDetails.webUrl, `is`(equalTo("webUrl")))
    }

    @Test
    fun testGetChangesHref() {
        `when`(build.changes).thenReturn(changes)
        `when`(changes.href).thenReturn("href")
        assertThat(mBuildDetails.changesHref, `is`(equalTo("href")))
    }

    @Test
    fun testGetStatusTextIfBuildIsNotQueued() {
        `when`(build.statusText).thenReturn("status")
        `when`(build.state).thenReturn("state")
        assertThat(mBuildDetails.statusText, `is`(equalTo("status")))
    }

    @Test
    fun testGetStatusTextIfBuildIsQueuedButDoesNotHaveWaitReason() {
        `when`(build.state).thenReturn(STATE_QUEUED)
        assertThat(mBuildDetails.statusText, `is`(equalTo(TEXT_QUEUED_BUILD)))
    }

    @Test
    fun testGetStatusTextIfBuildIsQueuedButHasWaitReason() {
        `when`(build.state).thenReturn(STATE_QUEUED)
        `when`(build.waitReason).thenReturn("waitReason")
        assertThat(mBuildDetails.statusText, `is`(equalTo("waitReason")))
    }

    @Test
    fun testGetStatusIcon() {
        `when`(build.state).thenReturn("")
        `when`(build.status).thenReturn("")
        assertThat(mBuildDetails.statusIcon, `is`(equalTo(IconUtils.ICON_SUCCESS)))
    }

    @Test
    fun testIsRunning() {
        `when`(build.state).thenReturn(STATE_RUNNING)
        assertThat(mBuildDetails.isRunning, `is`(equalTo(true)))
    }

    @Test
    fun testIsQueued() {
        `when`(build.state).thenReturn(STATE_QUEUED)
        assertThat(mBuildDetails.isQueued, `is`(equalTo(true)))
    }

    @Test
    fun testIsSuccess() {
        `when`(build.status).thenReturn(STATUS_SUCCESS)
        assertThat(mBuildDetails.isSuccess, `is`(equalTo(true)))
    }

    @Test
    fun testIsFailed() {
        `when`(build.status).thenReturn(STATUS_FAILURE)
        assertThat(mBuildDetails.isFailed, `is`(equalTo(true)))
    }

    @Test
    fun testHasCancellationInfo() {
        `when`(build.canceledInfo).thenReturn(canceledInfo)
        assertThat(mBuildDetails.hasCancellationInfo(), `is`(equalTo(true)))
    }

    @Test
    fun testHasUserInfoWhoCancelledBuild() {
        `when`(build.canceledInfo).thenReturn(canceledInfo)
        `when`(canceledInfo.user).thenReturn(user)
        assertThat(mBuildDetails.hasUserInfoWhoCancelledBuild(), `is`(equalTo(true)))
    }

    @Test
    fun testGetUserNameWhoCancelledBuild() {
        `when`(build.canceledInfo).thenReturn(canceledInfo)
        `when`(canceledInfo.user).thenReturn(user)
        `when`(user.name).thenReturn("user")
        assertThat(mBuildDetails.userNameWhoCancelledBuild, `is`(equalTo("user")))
    }

    @Test
    fun testGetCancellationTime() {
        `when`(build.canceledInfo).thenReturn(canceledInfo)
        `when`(canceledInfo.timestamp).thenReturn(TIME_STAMP)
        assertThat(mBuildDetails.cancellationTime, `is`(equalTo(TIME_FORMATTED)))
    }

    @Test
    fun testGetStartDate() {
        `when`(build.startDate).thenReturn(TIME_STAMP)
        assertThat(mBuildDetails.startDate, `is`(equalTo(TIME_FORMATTED)))
    }

    @Test
    fun testGetStartDateFormattedAsHeader() {
        `when`(build.startDate).thenReturn(TIME_STAMP)
        assertThat(mBuildDetails.startDateFormattedAsHeader, `is`(equalTo("21 June")))
    }

    @Test
    fun testGetQueuedDate() {
        `when`(build.queuedDate).thenReturn(TIME_STAMP)
        assertThat(mBuildDetails.queuedDate, `is`(equalTo(TIME_FORMATTED)))
    }

    @Test
    fun testGetFinishTime() {
        `when`(build.startDate).thenReturn(TIME_STAMP)
        `when`(build.finishDate).thenReturn(TIME_STAMP)
        assertThat(mBuildDetails.finishTime, `is`(equalTo("21 Jun 16 23:30 - 23:30 (0s)")))
    }

    @Test
    fun testGetEstimatedStartTime() {
        `when`(build.startEstimate).thenReturn(TIME_STAMP)
        assertThat(mBuildDetails.estimatedStartTime, `is`(equalTo(TIME_FORMATTED)))
    }

    @Test
    fun testGetBranchName() {
        `when`(build.branchName).thenReturn("branch")
        assertThat(mBuildDetails.branchName, `is`(equalTo("branch")))
    }

    @Test
    fun testHasAgentInfo() {
        `when`(build.agent).thenReturn(mAgent)
        assertThat(mBuildDetails.hasAgentInfo(), `is`(equalTo(true)))
    }

    @Test
    fun testGetAgentName() {
        `when`(build.agent).thenReturn(mAgent)
        assertThat(mBuildDetails.agentName, `is`(equalTo("name")))
    }

    @Test
    fun testIsTriggeredByVcs() {
        `when`(build.triggered).thenReturn(triggered)
        `when`(triggered.isVcs).thenReturn(true)
        assertThat(mBuildDetails.isTriggeredByVcs, `is`(equalTo(true)))
    }

    @Test
    fun testIsTriggeredByUser() {
        `when`(build.triggered).thenReturn(triggered)
        `when`(triggered.isUser).thenReturn(true)
        assertThat(mBuildDetails.isTriggeredByUser, `is`(equalTo(true)))
    }

    @Test
    fun testIsRestarted() {
        `when`(build.triggered).thenReturn(triggered)
        `when`(triggered.isRestarted).thenReturn(true)
        assertThat(mBuildDetails.isRestarted, `is`(equalTo(true)))
    }

    @Test
    fun testIsTriggeredByBuildType() {
        `when`(build.triggered).thenReturn(triggered)
        `when`(triggered.isBuildType).thenReturn(true)
        assertThat(mBuildDetails.isTriggeredByBuildType, `is`(equalTo(true)))
    }

    @Test
    fun testIsTriggeredByUnknown() {
        `when`(build.triggered).thenReturn(triggered)
        `when`(triggered.isUnknown).thenReturn(true)
        assertThat(mBuildDetails.isTriggeredByUnknown, `is`(equalTo(true)))
    }

    @Test
    fun tetsGetTriggeredDetails() {
        `when`(build.triggered).thenReturn(triggered)
        `when`(triggered.details).thenReturn("details")
        assertThat(mBuildDetails.triggeredDetails, `is`(equalTo("details")))
    }

    @Test
    fun testGetUserNameOfUserWhoTriggeredBuild() {
        `when`(build.triggered).thenReturn(triggered)
        `when`(triggered.user).thenReturn(user)
        `when`(user.name).thenReturn("name")
        assertThat(mBuildDetails.userNameOfUserWhoTriggeredBuild, `is`(equalTo("name")))
    }

    @Test
    fun testGetNameOfTriggeredBuildTypeIfBuildTypeIsDeleted() {
        `when`(build.triggered).thenReturn(triggered)
        `when`(triggered.buildType).thenReturn(null)
        assertThat(mBuildDetails.nameOfTriggeredBuildType, `is`(equalTo(BuildDetailsImpl.TEXT_DELETED_CONFIGURATION)))
    }

    @Test
    fun testGetNameOfTriggeredBuildType() {
        `when`(build.triggered).thenReturn(triggered)
        `when`(triggered.buildType).thenReturn(buildType)
        `when`(buildType.projectName).thenReturn("pn")
        `when`(buildType.name).thenReturn("cn")
        assertThat(mBuildDetails.nameOfTriggeredBuildType, `is`(equalTo("pn cn")))
    }

    @Test
    fun testGetBuildTypeId() {
        `when`(build.buildTypeId).thenReturn("id")
        assertThat(mBuildDetails.buildTypeId, `is`(equalTo("id")))
    }

    @Test
    fun testHasBuildTypeInfo() {
        `when`(build.buildType).thenReturn(null)
        assertThat(mBuildDetails.hasBuildTypeInfo(), `is`(equalTo(false)))
    }

    @Test
    fun testGetBuildTypeFullName() {
        `when`(build.buildType).thenReturn(buildType)
        `when`(buildType.projectName).thenReturn("project")
        `when`(buildType.name).thenReturn("name")
        assertThat(mBuildDetails.buildTypeFullName, `is`(equalTo("project - name")))
    }

    @Test
    fun testGetBuildTypeName() {
        `when`(build.buildType).thenReturn(buildType)
        `when`(buildType.name).thenReturn("name")
        assertThat(mBuildDetails.buildTypeName, `is`(equalTo("name")))
    }

    @Test
    fun testGetProjectName() {
        `when`(build.buildType).thenReturn(buildType)
        `when`(buildType.projectName).thenReturn("project")
        assertThat(mBuildDetails.projectName, `is`(equalTo("project")))
    }

    @Test
    fun testGetProjectId() {
        `when`(build.buildType).thenReturn(buildType)
        `when`(buildType.projectId).thenReturn("id")
        assertThat(mBuildDetails.projectId, `is`(equalTo("id")))
    }

    @Test
    fun testGetBuildId() {
        `when`(build.getId()).thenReturn("id")
        assertThat(mBuildDetails.id, `is`(equalTo("id")))
    }

    @Test
    fun testGetProperties() {
        `when`(build.properties).thenReturn(properties)
        assertThat(mBuildDetails.properties, `is`(equalTo(properties)))
    }

    @Test
    fun testIsTriggeredByUserIfTriggeredIsNull() {
        `when`(build.triggered).thenReturn(null)
        assertThat(mBuildDetails.isTriggeredByUser("user"), `is`(equalTo(false)))
    }

    @Test
    fun testIsTriggeredByUserIfTriggeredByNotUser() {
        `when`(build.triggered).thenReturn(triggered)
        `when`(triggered.isUser).thenReturn(false)
        assertThat(mBuildDetails.isTriggeredByUser("user"), `is`(equalTo(false)))
    }

    @Test
    fun testIsTriggeredByUserIfUserIsNull() {
        `when`(build.triggered).thenReturn(triggered)
        `when`(triggered.isUser).thenReturn(true)
        `when`(triggered.user).thenReturn(null)
        assertThat(mBuildDetails.isTriggeredByUser("user"), `is`(equalTo(false)))
    }

    @Test
    fun testIsTriggeredByUserIfUserIsNotNull() {
        `when`(build.triggered).thenReturn(triggered)
        `when`(triggered.isUser).thenReturn(true)
        `when`(triggered.user).thenReturn(User("user", null))
        assertThat(mBuildDetails.isTriggeredByUser("user"), `is`(equalTo(true)))
    }

    @Test
    fun testHasTests() {
        `when`(build.testOccurrences).thenReturn(testOccurrences)
        assertThat(mBuildDetails.hasTests(), `is`(equalTo(true)))
    }

    @Test
    fun testGetTestsHref() {
        `when`(build.testOccurrences).thenReturn(testOccurrences)
        `when`(testOccurrences.href).thenReturn("href")
        assertThat(mBuildDetails.testsHref, `is`(equalTo("href")))
    }

    @Test
    fun testGetPassedTestsCount() {
        `when`(build.testOccurrences).thenReturn(testOccurrences)
        `when`(testOccurrences.passed).thenReturn(3)
        assertThat(mBuildDetails.passedTestCount, `is`(equalTo(3)))
    }

    @Test
    fun testGetFailedTestsCount() {
        `when`(build.testOccurrences).thenReturn(testOccurrences)
        `when`(testOccurrences.failed).thenReturn(4)
        assertThat(mBuildDetails.failedTestCount, `is`(equalTo(4)))
    }

    @Test
    fun testGetIgnoredTestsCount() {
        `when`(build.testOccurrences).thenReturn(testOccurrences)
        `when`(testOccurrences.ignored).thenReturn(5)
        assertThat(mBuildDetails.ignoredTestCount, `is`(equalTo(5)))
    }

    @Test
    fun testGetArtifactsHref() {
        `when`(build.artifacts).thenReturn(artifacts)
        `when`(artifacts.href).thenReturn("href")
        assertThat(mBuildDetails.artifactsHref, `is`(equalTo("href")))
    }

    @Test
    fun testGetNumber() {
        `when`(build.number).thenReturn("Number")
        assertThat(mBuildDetails.number, `is`(equalTo("Number")))
    }

    @Test
    fun testGetNumberIfNumberIsNull() {
        `when`(build.number).thenReturn(null)
        assertThat(mBuildDetails.number, `is`(equalTo(TEXT_NO_NUMBER)))
    }

    @Test
    fun testIsPersonal() {
        `when`(build.isPersonal).thenReturn(true)
        assertThat(mBuildDetails.isPersonal, `is`(equalTo(true)))
    }

    @Test
    fun testIsPinned() {
        `when`(build.isPinned).thenReturn(true)
        assertThat(mBuildDetails.isPinned, `is`(equalTo(true)))
    }

    @Test
    fun testToBuild() {
        assertThat(mBuildDetails.toBuild(), `is`(equalTo(build)))
    }
}
